packages/flame/lib/src/text/common/glyph.dart

```diff
@@ -0,0 +1,104 @@
+import 'package:flame/src/text/common/glyph_data.dart';
+import 'package:flame/src/text/common/sprite_font.dart';
+import 'package:meta/meta.dart';
+
+/// [Glyph] describes a single character/ligature/icon within a [SpriteFont].
+///
+/// A glyph has an associated "glyph box", which is the box where the glyph is
+/// logically located. Here "logically" means that it includes not only the
+/// character itself, but also some padding around it as necessary for the
+/// character's to look nice within the text. For all glyphs in a font, their
+/// boxes will have the same height (which is the font size), the same ascent
+/// and descent, but possibly different widths.
+///
+/// The properties [left], [top], [width] and [height] describe the location of
+/// the glyph box within the source image.
+///
+/// In addition to the logical "glyph box", a glyph may also have a "source
+/// box", which describes a rectangle within the source image where the glyph's
+/// pixels are actually located. The source box may be larger or smaller than
+/// the glyph box. It will be larger if the glyph has a particularly large
+/// flourish that trespasses upon other character's space; or smaller if the
+/// characters are packed too tightly in the source image, or if you're trying
+/// to improve rendering performance by not copying empty pixels.
+class Glyph {
+  Glyph(
+    this.char, {
+    required this.left,
+    required this.top,
+    double? width,
+    double? height,
+    double? srcLeft,
+    double? srcTop,
+    double? srcRight,
+    double? srcBottom,
```

Comment:
Can't these be non-nullable and default to -1 and then the assert just look on the -1?

Reply:
It wouldn't simplify the assert, since we'd still need to check that either all four are -1 or all four are >= 0.

Also, from the semantic perspective, if a property is nullable then it sends a clear message that the property can be omitted. However, if the property has the default value of `-1` then I feel the user may want to know what does that mean. However, the value of `-1` is temporary, its only purpose is to store "null" value in a non-nullable field, it will be replaced by a real value once the `Glyph` is sent into the `SpriteFont`.

Reply:
Fair enough!